Document discovery (#127)
discord-userdoccers/discord-userdoccers: pages/resources/discovery.mdx

# Discovery

### Searching Discovery

Discord utilizes [Algolia](https://www.algolia.com/) to power search for discovery. You can search for guilds by name, description, keywords, and more. Reference the [Algolia Search documentation](https://www.algolia.com/doc/rest-api/search/) for more information on how to search. A [proxied version of the Algolia Search API](#search-discoverable-guilds) is available, with limited parameter support.

<Alert type="info">

> All hits returned from Algolia are [discoverable guild](#discoverable-guild-object) objects. However, some fields such as `features` are truncated to only include values that are relevant to discovery.

Reviewer: Specify the features that are relevant if known pls

Author: I don't think its worth maintaining a list for that.

</Alert>

###### Algolia Credentials

> These credentials are for production only.

Reviewer: ```suggestion
These credentials are for Discord production’s Algolia client only. 
```

Author: Is it a client? I think that terminology is off.

| approximate_member_count   | integer                                                       | Approximate number of members in the guild                                                       |
| approximate_presence_count | integer                                                       | Approximate number of online members in the guild                                                |
| emojis? ^1^                | array[[emoji](/resources/emoji#emoji-object) object]          | Custom guild emojis; limited to 30 entries                                                       |
| emoji_count? ^1^           | integer                                                       | Total number of custom guild emojis                                                              |
| stickers? ^1^              | array[[sticker](/resources/sticker#sticker-object) object]    | Custom guild stickers; limited to 30 entries                                                     |
| sticker_count? ^1^         | integer                                                       | Total number of custom guild stickers                                                            |

> | auto_removed               | boolean                                                       | Whether the guild has automatically been removed from discovery for not hitting required targets |

Reviewer: Why is this in a discoverable guild object? Shouldn’t guilds with this be removed from Algolia automatically?

Author: I didn't design the API 😭
I don't think it's removed that quickly, and the client filters out auto-removed guilds when searching: `query=test&filters=auto_removed%3Afalse+AND+approximate_presence_count%3E+0+AND+approximate_member_count%3E200&optionalFilters=%5B%22preferred_locale%3A+en-US%22%5D&length=12&offset=0&restrictSearchableAttributes=%5B%22name%22%2C%22description%22%2C%22keywords%22%2C%22categories.name%22%2C%22categories.name_localizations.en-US%22%2C%22primary_category.name%22%2C%22primary_category.name_localizations.en-US%22%2C%22vanity_url_code%22%5D`. Also, there's a grace period before the guild is removed.

  "icon": "278da1c7740e394657c1179f4782aef1",
  "splash": "2b4ae5cdd71038b4880b1b57a6e5dacb",
  "banner": "57e939e67aebaf232d28205603020b55",
  "approximate_presence_count": 21125,
  "approximate_member_count": 155455,
  "premium_subscription_count": 17,
  "preferred_locale": "en-US",
  "auto_removed": false,
  "discovery_splash": "9d7ec672b89b320ef7a51e5b6ae453b8",
  "primary_category_id": 1,
  "vanity_url_code": "creative",
  "is_published": true,
  "keywords": [
    "Fortnite",
    "Creative",
    "Fortnite Creative",
    "Boxfights",
    "Zonewars",
    "Buildfights",
    "1v1",
    "EU",
    "NA",
    "Creator"
  ],
  "features": [
    "ANIMATED_BANNER",
    "ANIMATED_ICON",
    "AUTO_MODERATION",
    "BANNER",
    "COMMUNITY",
    "DISCOVERABLE",
    "ENABLED_DISCOVERABLE_BEFORE",
    "GUILD_ONBOARDING_EVER_ENABLED",
    "GUILD_WEB_PAGE_VANITY_URL",
    "INVITE_SPLASH",
    "NEWS",
    "PREVIEW_ENABLED",
    "RAID_ALERTS_ENABLED",
    "ROLE_ICONS",
    "VANITY_URL",
    "WELCOME_SCREEN_ENABLED"
  ],
  "emojis": [],
  "emoji_count": 339
}
```

### Discovery Requirements Object

> Represents a guild's path towards meeting the requirements of joining discovery.

Reviewer: ```suggestion
Represents a guild's progress on meeting the requirements of joining discovery.
```

| minimum_size?                       | ?integer                                                                 | The minimum guild member count requirement                                                             |
| grace_period_end_date?              | ISO8601 timestamp                                                        | When the guild's grace period ends                                                                     |

^1^ Certain guilds, such as those that are [verified](https://discord.com/verification), are exempt from discovery requirements. These guilds will not have a fully populated discovery requirements object (guaranteed to receive only `sufficient` and `sufficient_without_grace_period`).

###### Discovery NSFW Properties Structure

| Field                        | Type                          | Description                                                    |
| ---------------------------- | ----------------------------- | -------------------------------------------------------------- |

> | channels?                    | array[snowflake]              | The IDs of the channels with names containing disallowed terms |

Reviewer: ```suggestion
| channels?                    | array[snowflake]              | The channels with names containing disallowed terms |
```

Author: This would introduce inconsistency.

###### Discovery Reason Structure

| Field      | Type       | Description                                                       |
| ---------- | ---------- | ----------------------------------------------------------------- |
| reason     | string     | The reason to join the guild                                      |
| emoji_id   | ?snowflake | The [ID of a guild's custom emoji](/resources/emoji#emoji-object) |
| emoji_name | ?string    | The unicode character of the emoji                                |

###### Example Discovery Metadata

```json
{
  "guild_id": "1029315212005888060",
  "primary_category_id": 49,
  "keywords": ["test"],
  "emoji_discoverability_enabled": true,
  "partner_actioned_timestamp": null,
  "partner_application_timestamp": null,
  "is_published": false,
  "reasons_to_join": [
    { "reason": "Alien", "emoji_id": null, "emoji_name": "👽" },
    { "reason": "Alien", "emoji_id": null, "emoji_name": "👽" },
    { "reason": "Alien", "emoji_id": null, "emoji_name": "👽" },

>     { "reason": "Alien", "emoji_id": null, "emoji_name": "👽" }

Reviewer: Include a custom emoji here just for example completeness 😛
